File: CSharp_Code/MySaveDialogControl.cs
//  Copyright (c) 2006, Gustavo Franco
//  Copyright © Decebal Mihailescu 2007-2015

//  Email:  gustavo_franco@hotmail.com
//  All rights reserved.

//  Redistribution and use in source and binary forms, with or without modification, 
//  are permitted provided that the following conditions are met:

//  Redistributions of source code must retain the above copyright notice, 
//  this list of conditions and the following disclaimer. 
//  Redistributions in binary form must reproduce the above copyright notice, 
//  this list of conditions and the following disclaimer in the documentation 
//  and/or other materials provided with the distribution. 

//  THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
//  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR
//  PURPOSE. IT CAN BE DISTRIBUTED FREE OF CHARGE AS LONG AS THIS HEADER 
//  REMAINS UNCHANGED.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Drawing.Imaging;
using System.IO;
using FileDialogExtenders;


namespace CustomControls
{
    public partial class MySaveDialogControl : FileDialogControlBase
    {
        Bitmap _changedImage;
        internal Image _originalImage;
        RotateFlipType _rft;
        bool _canDisplay = true;
        ImageFormat _format = ImageFormat.Jpeg;
        System.IO.MemoryStream _memstream;
        string _imageFile;
        public MySaveDialogControl()
        {
            InitializeComponent();
        }

        protected override void OnPrepareMSDialog()
        {
            base.OnPrepareMSDialog();
            MSDialog.FilterIndex = GetIndexfromFile(_imageFile);
            if (Environment.OSVersion.Version.Major < 6)
                MSDialog.SetPlaces(new object[] { (int)Places.Desktop, (int)Places.Printers, (int)Places.Favorites, (int)Places.Programs, (int)Places.Fonts, });
            FileDlgInitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures);
        }

        public MySaveDialogControl(string originalfile, IWin32Window parent)
        {
            try
            {
                _imageFile = originalfile;
                _originalImage = Image.FromFile(originalfile);                
            }
            catch 
            {
                _imageFile = null;
            }

            if (_imageFile == null || _imageFile.Length == 0)
            {

                MyOpenFileDialogControl openDialog = new MyOpenFileDialogControl();
                try
                {
                    if (parent == null)
                        parent = Application.OpenForms[0];
                    if (openDialog.ShowDialog(parent) == DialogResult.OK)
                    {
                        _imageFile = openDialog.MSDialog.FileName;
                        _originalImage =  Image.FromFile(_imageFile);
                        InitializeComponent();
                    }
                    else
                    {
                        Dispose();
                    }
                }
                finally
                {
                    if (openDialog != null)
                        openDialog.Dispose();
                    openDialog = null;
                    GC.Collect();
                    GC.WaitForPendingFinalizers();
                }
            }
            else
            {
                
                InitializeComponent();
            }
            
        }
        private void MySaveDialogControl_Load(object sender, EventArgs e)
        {
            if (!DesignMode)
            {
                FileInfo fi = new FileInfo(_imageFile);
                _lblOriginalFileLen.Text = string.Format("Original File Size: {0} bytes", fi.Length);
            }
            _horizontal.Minimum = _pbChanged.MinimumSize.Width;
            _horizontal.Maximum = _pbChanged.MaximumSize.Width/2;   
            _vertical.Minimum = _pbChanged.MinimumSize.Height;
            _vertical.Maximum = _pbChanged.MaximumSize.Height/2; 
            _horizontal.Value = _pbChanged.MaximumSize.Width/2;
            _vertical.Value = _pbChanged.MaximumSize.Height/2;         
            _groupBox2.Text = System.IO.Path.GetFileName(_imageFile);
            _groupBox1.Text = "New Image";
            _rotateflip.DataSource = Enum.GetNames(typeof(RotateFlipType));
            _rotateFlipOriginal.DataSource = Enum.GetNames(typeof(RotateFlipType));
            _cbxOriginalMode.DataSource = Enum.GetNames(typeof(System.Windows.Forms.PictureBoxSizeMode));
            _cbxNewViewMode.DataSource = Enum.GetNames(typeof(System.Windows.Forms.PictureBoxSizeMode));
            _cbxNewViewMode.SelectedIndex = (int)PictureBoxSizeMode.Zoom;
            _cbxOriginalMode.SelectedIndex = (int)PictureBoxSizeMode.Zoom;
            _pbChanged.SizeMode = PictureBoxSizeMode.Zoom;
            _pbOriginal.SizeMode = PictureBoxSizeMode.Zoom;
            _pbChanged.MaximumSize = _pbChanged.Size;
            _pbOriginal.MaximumSize = _pbOriginal.Size;
            if (_originalImage != null && !DesignMode)
            {
                _pbOriginal.Image = _originalImage;
                _format = _originalImage.RawFormat;
                _lblOrigInfo.Text = string.Format("Colors: {0}\nSize: {1} X {2} pixels\nHorizontal Resolution {3}\nVertical resolution: {4}",
                    MyOpenFileDialogControl.GetColorsCountFromImage(_pbOriginal.Image),_originalImage.Width,
                    _originalImage.Height,_originalImage.HorizontalResolution, _originalImage.VerticalResolution);
                Display();
            }
        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        protected override void Dispose(bool disposing)
        {
            if (_memstream != null)
            {
                _memstream.Dispose();
                _memstream = null;
            }
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        private void _rotateflip_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (_changedImage == null)
                return;
            if (sender == _rotateflip)
            {
                _rft = (RotateFlipType)Enum.Parse(typeof(RotateFlipType), _rotateflip.SelectedItem.ToString());
                Display();
            }
            else
            {
                _pbOriginal.Image.RotateFlip((RotateFlipType)Enum.Parse(typeof(RotateFlipType), _rotateFlipOriginal.SelectedItem.ToString()));
                _pbOriginal.Refresh();
            }


        }

        private void _horizontal_ValueChanged(object sender, EventArgs e)
        {
            if (_originalImage == null || !_canDisplay)
                return;
            if (_ckbPreserveaspect.Checked)
                _vertical.Value = _horizontal.Value;
        }


        private void _vertical_ValueChanged(object sender, EventArgs e)
        {
            if (_originalImage == null || !_canDisplay)
                return;
        }

        private void _Update_Click(object sender, EventArgs e)
        {
            Display();
        }

        bool Display()
        {
            try
            {
                if (_originalImage == null && _canDisplay)
                    return false;
                double aspectfactor = (double)_originalImage.Height / _originalImage.Width;
                int verticalVal =  (int)this._vertical.Value ;
                int horizontalVal = (int)this._horizontal.Value ;
                if (aspectfactor > 1)
                {//portrait
                    horizontalVal = (int)(horizontalVal / aspectfactor);
                }
                else
                {//landscape
                    verticalVal = (int)(verticalVal * aspectfactor);
                }
                Bitmap newImg = _originalImage.GetThumbnailImage(horizontalVal, verticalVal, null, IntPtr.Zero) as Bitmap;
                if (_changedImage != null)
                    _changedImage.Dispose();
                newImg.RotateFlip(_rft);
                _pbChanged.SuspendLayout();
                _pbOriginal.SuspendLayout();
                if (_cbxNewViewMode.SelectedItem == null)
                    _pbChanged.SizeMode = PictureBoxSizeMode.AutoSize;
                else
                    _pbChanged.SizeMode = (PictureBoxSizeMode)Enum.Parse(typeof(PictureBoxSizeMode), _cbxNewViewMode.SelectedItem.ToString());
                if (_cbxOriginalMode.SelectedItem == null)
                    _pbOriginal.SizeMode = PictureBoxSizeMode.AutoSize;
                else
                    _pbOriginal.SizeMode = (PictureBoxSizeMode)Enum.Parse(typeof(PictureBoxSizeMode), _cbxOriginalMode.SelectedItem.ToString());
                _pbChanged.Image = _changedImage = newImg;
                _pbChanged.Size = _pbChanged.MaximumSize;
                _pbChanged.ResumeLayout(true);
                _pbOriginal.ResumeLayout(true);
                return ShowImageInfo();
            }
            catch
            {
                return false;
            }
        }

        private bool ShowImageInfo()
        {
            try
            {
                if (_memstream != null)
                {
                    _memstream.Dispose();
                    _memstream = null;
                }
                _memstream = new MemoryStream();
                _changedImage.Save(_memstream, _format);
                //correct the format bug
                ImageFormat ifrm = default(ImageFormat);
                if (_format.Equals(ImageFormat.Bmp))
                    ifrm = ImageFormat.Bmp;
                else if (_format.Equals(ImageFormat.Gif))
                    ifrm = ImageFormat.Gif;
                else if (_format.Equals(ImageFormat.Jpeg))
                    ifrm = ImageFormat.Jpeg;
                else if (_format.Equals(ImageFormat.Png))
                    ifrm = ImageFormat.Png;
                else if (_format.Equals(ImageFormat.Tiff))
                    ifrm = ImageFormat.Tiff;
                else if (_format.Equals(ImageFormat.Emf))
                    ifrm = ImageFormat.Emf;
                else if (_format.Equals(ImageFormat.Icon))
                    ifrm = ImageFormat.Icon;
                else if (_format.Equals(ImageFormat.Wmf))
                    ifrm = ImageFormat.Wmf;
                else if (_format.Equals(ImageFormat.Exif))
                    ifrm = ImageFormat.Exif;


                _FileSize.Text = string.Format("Picture Size: {0} bytes for {3}\nWidth = {1} Height ={2} pixels",
                    _memstream.Length, (int)_horizontal.Value, (int)_vertical.Value, ifrm.ToString());
                return true;
            }
            catch 
            {
                throw;
            }
        }

        private void UPDOWN_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left && sender is UpDownBase)
                _canDisplay = false;
            if (_ckbPreserveaspect.Checked)
                _vertical.Value = _horizontal.Value;

        }

        private void UPDOWN_MouseUp(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left && sender is UpDownBase)
            if (_ckbPreserveaspect.Checked)
                _vertical.Value = _horizontal.Value;
                _canDisplay = true;
            Display();
        }

        private void MySaveDialogControl_FilterChanged(IWin32Window sender, int index)
        {
            FileDlgEnableOkBtn = GetFormatFromIndex(index);
        }

        private bool GetFormatFromIndex(int index)
        {
            string ext = string.Empty;
            ImageFormat oldform = _format;
            try
            {
                string[] extensions = MSDialog.Filter.Split('|');
                index--;
                ext = extensions[2 * index + 1];
                int pos = ext.LastIndexOf('.');
                ext = ext.Substring(pos + 1).Trim().ToLower();
                switch (ext)
                {
                    case "jpg":
                    case "jpeg":
                        _format = ImageFormat.Jpeg;
                        break;
                    case "bmp":
                        _format = ImageFormat.Bmp;
                        break;
                    case "gif":
                        _format = ImageFormat.Gif;
                        break;
                    case "emf":
                        _format = ImageFormat.Emf;
                        break;
                    case "ico":
                        _format = ImageFormat.Icon;
                        break;
                    case "png":
                        _format = ImageFormat.Png;
                        break;
                    case "tif":
                        _format = ImageFormat.Tiff;
                        break;
                    case "wmf":
                        _format = ImageFormat.Wmf;
                        break;
                    case "exif":
                        _format = ImageFormat.Exif;
                        break;
                    default:
                        throw new ApplicationException(ext + " is an unknown format.Select a known one.");
                }
                return ShowImageInfo();
            }
            catch (Exception ex)
            {
                MessageBox.Show(this, ex.Message, "Conversion error for ." +ext+ " type",
                    MessageBoxButtons.OK, MessageBoxIcon.Error);
                //revert to last good format
                MSDialog.FilterIndex = GetIndexfromFile("dummyfile." + oldform.ToString());
                _format = oldform;
                return false;
            }         
        }

       private int GetIndexfromFile(string file)
        {

            try
            {
            string ext = System.IO.Path.GetExtension(file).ToLower();
            int index = MSDialog.Filter.ToLower().IndexOf(ext);
            string[] extensions = MSDialog.Filter.Substring(0,index).Split(new char[] { '|' }, StringSplitOptions.RemoveEmptyEntries);
            return extensions.Length / 2 + 1;
            }
            catch (Exception ex)
            {
                MessageBox.Show(this, ex.Message, "Unknown Format!", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return 1;
            }         
        }
        private void MySaveDialogControl_ClosingDialog(object sender, CancelEventArgs e)
        {
            if (_memstream == null)
                return;
            System.IO.FileStream br = null;
            try
            {

                    br = new FileStream(MSDialog.FileName, FileMode.OpenOrCreate, FileAccess.Write);
                    _memstream.WriteTo(br);
                    br.Flush();
                    br.Close();
                    e.Cancel = false;
            }
            catch(Exception ex)
            {
                MessageBox.Show(this, "Unable to save the file because:\n" 
                    + ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Hand);
                e.Cancel = true;
            }
            finally
            {
                if(br != null)
                    br.Dispose();
                _memstream.Dispose();
                _memstream = null;
            }

        }


        private void ckbOrigStretch_CheckedChanged(object sender, EventArgs e)
        {
            Display();
        }

        private void ckbPreserveaspect_CheckedChanged(object sender, EventArgs e)
        {
            _vertical.Enabled = !_ckbPreserveaspect.Checked;
            if (_ckbPreserveaspect.Checked)
                _vertical.Value = _horizontal.Value;
        }

        private void cbxNewViewMode_SelectedIndexChanged(object sender, EventArgs e)
        {
            if(sender == _cbxNewViewMode)
                _pbChanged.SizeMode = (PictureBoxSizeMode)Enum.Parse(typeof(PictureBoxSizeMode), _cbxNewViewMode.SelectedItem.ToString());
            else
                _pbOriginal.SizeMode = (PictureBoxSizeMode)Enum.Parse(typeof(PictureBoxSizeMode), _cbxOriginalMode.SelectedItem.ToString());
            Display();
        }

        private void MySaveDialogControl_HelpRequested(object sender, HelpEventArgs hlpevent)
        {
            MessageBox.Show("Please add some specific help here");
        }

    }
}

File: VB_Code/MySaveDialogControl.vb
'  Copyright (c) 2006, Gustavo Franco
'  Copyright © Decebal Mihailescu 2007-2015

'  Email:  gustavo_franco@hotmail.com
'  All rights reserved.

'  Redistribution and use in source and binary forms, with or without modification, 
'  are permitted provided that the following conditions are met:

'  Redistributions of source code must retain the above copyright notice, 
'  this list of conditions and the following disclaimer. 
'  Redistributions in binary form must reproduce the above copyright notice, 
'  this list of conditions and the following disclaimer in the documentation 
'  and/or other materials provided with the distribution. 

'  THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR
'  PURPOSE. IT CAN BE DISTRIBUTED FREE OF CHARGE AS LONG AS THIS HEADER 
'  REMAINS UNCHANGED.


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing.Imaging
Imports System.IO
Imports FileDialogExtenders


Namespace CustomControls
	Partial Public Class MySaveDialogControl
		Inherits FileDialogControlBase
		Private _changedImage As Bitmap
		Friend _originalImage As Image
		Private _rft As RotateFlipType
		Private _canDisplay As Boolean = True
		Private _format As ImageFormat = ImageFormat.Jpeg
		Private _memstream As System.IO.MemoryStream
		Private _imageFile As String
		Public Sub New()
			InitializeComponent()
		End Sub

		Protected Overrides Sub OnPrepareMSDialog()
			MyBase.OnPrepareMSDialog()
			MSDialog.FilterIndex = GetIndexfromFile(_imageFile)
			If Environment.OSVersion.Version.Major < 6 Then
				MSDialog.SetPlaces(New Object() { CInt(Places.Desktop), CInt(Places.Printers), CInt(Places.Favorites), CInt(Places.Programs), CInt(Places.Fonts) })
			End If
			FileDlgInitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)
		End Sub

		Public Sub New(ByVal originalfile As String, ByVal parent As IWin32Window)
			Try
				_imageFile = originalfile
				_originalImage = Image.FromFile(originalfile)
			Catch
				_imageFile = Nothing
			End Try

			If _imageFile Is Nothing OrElse _imageFile.Length = 0 Then

				Dim openDialog As New MyOpenFileDialogControl()
				Try
					If parent Is Nothing Then
						parent = Application.OpenForms(0)
					End If
					If openDialog.ShowDialog(parent) = DialogResult.OK Then
						_imageFile = openDialog.MSDialog.FileName
						_originalImage = Image.FromFile(_imageFile)
						InitializeComponent()
					Else
						Dispose()
					End If
				Finally
					If openDialog IsNot Nothing Then
						openDialog.Dispose()
					End If
					openDialog = Nothing
					GC.Collect()
					GC.WaitForPendingFinalizers()
				End Try
			Else

				InitializeComponent()
			End If

		End Sub
		Private Sub MySaveDialogControl_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			If Not DesignMode Then
				Dim fi As New FileInfo(_imageFile)
				_lblOriginalFileLen.Text = String.Format("Original File Size: {0} bytes", fi.Length)
			End If
			_horizontal.Minimum = _pbChanged.MinimumSize.Width
			_horizontal.Maximum = _pbChanged.MaximumSize.Width\2
			_vertical.Minimum = _pbChanged.MinimumSize.Height
			_vertical.Maximum = _pbChanged.MaximumSize.Height\2
			_horizontal.Value = _pbChanged.MaximumSize.Width\2
			_vertical.Value = _pbChanged.MaximumSize.Height\2
			_groupBox2.Text = System.IO.Path.GetFileName(_imageFile)
			_groupBox1.Text = "New Image"
			_rotateflip.DataSource = System.Enum.GetNames(GetType(RotateFlipType))
			_rotateFlipOriginal.DataSource = System.Enum.GetNames(GetType(RotateFlipType))
			_cbxOriginalMode.DataSource = System.Enum.GetNames(GetType(System.Windows.Forms.PictureBoxSizeMode))
			_cbxNewViewMode.DataSource = System.Enum.GetNames(GetType(System.Windows.Forms.PictureBoxSizeMode))
			_cbxNewViewMode.SelectedIndex = CInt(PictureBoxSizeMode.Zoom)
			_cbxOriginalMode.SelectedIndex = CInt(PictureBoxSizeMode.Zoom)
			_pbChanged.SizeMode = PictureBoxSizeMode.Zoom
			_pbOriginal.SizeMode = PictureBoxSizeMode.Zoom
			_pbChanged.MaximumSize = _pbChanged.Size
			_pbOriginal.MaximumSize = _pbOriginal.Size
			If _originalImage IsNot Nothing AndAlso (Not DesignMode) Then
				_pbOriginal.Image = _originalImage
				_format = _originalImage.RawFormat
				_lblOrigInfo.Text = String.Format("Colors: {0}" & Constants.vbLf & "Size: {1} X {2} pixels" & Constants.vbLf & "Horizontal Resolution {3}" & Constants.vbLf & "Vertical resolution: {4}", MyOpenFileDialogControl.GetColorsCountFromImage(_pbOriginal.Image),_originalImage.Width, _originalImage.Height,_originalImage.HorizontalResolution, _originalImage.VerticalResolution)
				Display()
			End If
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		''' <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If _memstream IsNot Nothing Then
				_memstream.Dispose()
				_memstream = Nothing
			End If
			If disposing AndAlso (components IsNot Nothing) Then
				components.Dispose()
			End If
			MyBase.Dispose(disposing)
		End Sub

		Private Sub _rotateflip_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _rotateflip.SelectedIndexChanged, _rotateFlipOriginal.SelectedIndexChanged
			If _changedImage Is Nothing Then
				Return
			End If
			If sender Is _rotateflip Then
				_rft = CType(System.Enum.Parse(GetType(RotateFlipType), _rotateflip.SelectedItem.ToString()), RotateFlipType)
				Display()
			Else
				_pbOriginal.Image.RotateFlip(CType(System.Enum.Parse(GetType(RotateFlipType), _rotateFlipOriginal.SelectedItem.ToString()), RotateFlipType))
				_pbOriginal.Refresh()
			End If


		End Sub

		Private Sub _horizontal_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _horizontal.ValueChanged
			If _originalImage Is Nothing OrElse (Not _canDisplay) Then
				Return
			End If
			If _ckbPreserveaspect.Checked Then
				_vertical.Value = _horizontal.Value
			End If
		End Sub


		Private Sub _vertical_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _vertical.ValueChanged
			If _originalImage Is Nothing OrElse (Not _canDisplay) Then
				Return
			End If
		End Sub

		Private Sub _Update_Click(ByVal sender As Object, ByVal e As EventArgs)
			Display()
		End Sub

		Private Function Display() As Boolean
			Try
				If _originalImage Is Nothing AndAlso _canDisplay Then
					Return False
				End If
				Dim aspectfactor As Double = CDbl(_originalImage.Height) / _originalImage.Width
				Dim verticalVal As Integer = CInt(Fix(Me._vertical.Value))
				Dim horizontalVal As Integer = CInt(Fix(Me._horizontal.Value))
				If aspectfactor > 1 Then
					horizontalVal = CInt(Fix(horizontalVal / aspectfactor))
				Else
					verticalVal = CInt(Fix(verticalVal * aspectfactor))
				End If
				Dim newImg As Bitmap = TryCast(_originalImage.GetThumbnailImage(horizontalVal, verticalVal, Nothing, IntPtr.Zero), Bitmap)
				If _changedImage IsNot Nothing Then
					_changedImage.Dispose()
				End If
				newImg.RotateFlip(_rft)
				_pbChanged.SuspendLayout()
				_pbOriginal.SuspendLayout()
				If _cbxNewViewMode.SelectedItem Is Nothing Then
					_pbChanged.SizeMode = PictureBoxSizeMode.AutoSize
				Else
					_pbChanged.SizeMode = CType(System.Enum.Parse(GetType(PictureBoxSizeMode), _cbxNewViewMode.SelectedItem.ToString()), PictureBoxSizeMode)
				End If
				If _cbxOriginalMode.SelectedItem Is Nothing Then
					_pbOriginal.SizeMode = PictureBoxSizeMode.AutoSize
				Else
					_pbOriginal.SizeMode = CType(System.Enum.Parse(GetType(PictureBoxSizeMode), _cbxOriginalMode.SelectedItem.ToString()), PictureBoxSizeMode)
				End If
				_changedImage = newImg
				_pbChanged.Image = _changedImage
				_pbChanged.Size = _pbChanged.MaximumSize
				_pbChanged.ResumeLayout(True)
				_pbOriginal.ResumeLayout(True)
				Return ShowImageInfo()
			Catch
				Return False
			End Try
		End Function

		Private Function ShowImageInfo() As Boolean
			Try
				If _memstream IsNot Nothing Then
					_memstream.Dispose()
					_memstream = Nothing
				End If
				_memstream = New MemoryStream()
				_changedImage.Save(_memstream, _format)
				'correct the format bug
				Dim ifrm As ImageFormat = Nothing
				If _format.Equals(ImageFormat.Bmp) Then
					ifrm = ImageFormat.Bmp
				ElseIf _format.Equals(ImageFormat.Gif) Then
					ifrm = ImageFormat.Gif
				ElseIf _format.Equals(ImageFormat.Jpeg) Then
					ifrm = ImageFormat.Jpeg
				ElseIf _format.Equals(ImageFormat.Png) Then
					ifrm = ImageFormat.Png
				ElseIf _format.Equals(ImageFormat.Tiff) Then
					ifrm = ImageFormat.Tiff
				ElseIf _format.Equals(ImageFormat.Emf) Then
					ifrm = ImageFormat.Emf
				ElseIf _format.Equals(ImageFormat.Icon) Then
					ifrm = ImageFormat.Icon
				ElseIf _format.Equals(ImageFormat.Wmf) Then
					ifrm = ImageFormat.Wmf
				ElseIf _format.Equals(ImageFormat.Exif) Then
					ifrm = ImageFormat.Exif
				End If


				_FileSize.Text = String.Format("Picture Size: {0} bytes for {3}" & Constants.vbLf & "Width = {1} Height ={2} pixels", _memstream.Length, CInt(Fix(_horizontal.Value)), CInt(Fix(_vertical.Value)), ifrm.ToString())
				Return True
			Catch
				Throw
			End Try
		End Function

		Private Sub UPDOWN_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles _horizontal.MouseDown, _vertical.MouseDown
			If e.Button = MouseButtons.Left AndAlso TypeOf sender Is UpDownBase Then
				_canDisplay = False
			End If
			If _ckbPreserveaspect.Checked Then
				_vertical.Value = _horizontal.Value
			End If

		End Sub

		Private Sub UPDOWN_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs) Handles _horizontal.MouseUp, _vertical.MouseUp
			If e.Button = MouseButtons.Left AndAlso TypeOf sender Is UpDownBase Then
			If _ckbPreserveaspect.Checked Then
				_vertical.Value = _horizontal.Value
			End If
			End If
				_canDisplay = True
			Display()
		End Sub

		Private Sub MySaveDialogControl_FilterChanged(ByVal sender As IWin32Window, ByVal index As Integer) Handles MyBase.EventFilterChanged
			FileDlgEnableOkBtn = GetFormatFromIndex(index)
		End Sub

		Private Function GetFormatFromIndex(ByVal index As Integer) As Boolean
			Dim ext As String = String.Empty
			Dim oldform As ImageFormat = _format
			Try
				Dim extensions() As String = MSDialog.Filter.Split("|"c)
				index -= 1
				ext = extensions(2 * index + 1)
				Dim pos As Integer = ext.LastIndexOf("."c)
				ext = ext.Substring(pos + 1).Trim().ToLower()
				Select Case ext
					Case "jpg", "jpeg"
						_format = ImageFormat.Jpeg
					Case "bmp"
						_format = ImageFormat.Bmp
					Case "gif"
						_format = ImageFormat.Gif
					Case "emf"
						_format = ImageFormat.Emf
					Case "ico"
						_format = ImageFormat.Icon
					Case "png"
						_format = ImageFormat.Png
					Case "tif"
						_format = ImageFormat.Tiff
					Case "wmf"
						_format = ImageFormat.Wmf
					Case "exif"
						_format = ImageFormat.Exif
					Case Else
						Throw New ApplicationException(ext & " is an unknown format.Select a known one.")
				End Select
				Return ShowImageInfo()
			Catch ex As Exception
				MessageBox.Show(Me, ex.Message, "Conversion error for ." & ext & " type", MessageBoxButtons.OK, MessageBoxIcon.Error)
				'revert to last good format
				MSDialog.FilterIndex = GetIndexfromFile("dummyfile." & oldform.ToString())
				_format = oldform
				Return False
			End Try
		End Function

	   Private Function GetIndexfromFile(ByVal file As String) As Integer

			Try
			Dim ext As String = System.IO.Path.GetExtension(file).ToLower()
			Dim index As Integer = MSDialog.Filter.ToLower().IndexOf(ext)
			Dim extensions() As String = MSDialog.Filter.Substring(0,index).Split(New Char() { "|"c }, StringSplitOptions.RemoveEmptyEntries)
			Return extensions.Length \ 2 + 1
			Catch ex As Exception
				MessageBox.Show(Me, ex.Message, "Unknown Format!", MessageBoxButtons.OK, MessageBoxIcon.Error)
				Return 1
			End Try
	   End Function
		Private Sub MySaveDialogControl_ClosingDialog(ByVal sender As Object, ByVal e As CancelEventArgs) Handles MyBase.EventClosingDialog
			If _memstream Is Nothing Then
				Return
			End If
			Dim br As System.IO.FileStream = Nothing
			Try

					br = New FileStream(MSDialog.FileName, FileMode.OpenOrCreate, FileAccess.Write)
					_memstream.WriteTo(br)
					br.Flush()
					br.Close()
					e.Cancel = False
			Catch ex As Exception
				MessageBox.Show(Me, "Unable to save the file because:" & Constants.vbLf + ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Hand)
				e.Cancel = True
			Finally
				If br IsNot Nothing Then
					br.Dispose()
				End If
				_memstream.Dispose()
				_memstream = Nothing
			End Try

		End Sub


		Private Sub ckbOrigStretch_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
			Display()
		End Sub

		Private Sub ckbPreserveaspect_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _ckbPreserveaspect.CheckedChanged
			_vertical.Enabled = Not _ckbPreserveaspect.Checked
			If _ckbPreserveaspect.Checked Then
				_vertical.Value = _horizontal.Value
			End If
		End Sub

		Private Sub cbxNewViewMode_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cbxNewViewMode.SelectedIndexChanged, _cbxOriginalMode.SelectedIndexChanged
			If sender Is _cbxNewViewMode Then
				_pbChanged.SizeMode = CType(System.Enum.Parse(GetType(PictureBoxSizeMode), _cbxNewViewMode.SelectedItem.ToString()), PictureBoxSizeMode)
			Else
				_pbOriginal.SizeMode = CType(System.Enum.Parse(GetType(PictureBoxSizeMode), _cbxOriginalMode.SelectedItem.ToString()), PictureBoxSizeMode)
			End If
			Display()
		End Sub

		Private Sub MySaveDialogControl_HelpRequested(ByVal sender As Object, ByVal hlpevent As HelpEventArgs) Handles MyBase.HelpRequested
			MessageBox.Show("Please add some specific help here")
		End Sub

	End Class
End Namespace

